core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -451,6 +451,15 @@ protected void doRun() {
                 }
                 performAction(node, transportPrimaryAction, true);
             } else {
+                if (state.version() < request.minimumClusterStateVersion()) {
+                    logger.trace("cluster state [{}] on node [{}] should have minimum version [{}] but was [{}], scheduling a retry: action [{}], request [{}]", request.shardId(), primary.currentNodeId(), request.minimumClusterStateVersion(), state.version(), actionName, request);
+                    retryBecauseUnavailable(request.shardId(), "expected minimum cluster state version " + request.minimumClusterStateVersion() + " but was " + state.version());
+                    return;
+                } else {
+                    // chasing the node with the active primary for a second hop requires that we are at least up-to-date with the current cluster state version
+                    // this prevents redirect loops between two nodes when a primary was relocated and the relocation target is not aware that it is the active primary shard already.
+                    request.minimumClusterStateVersion(state.version());
```

Comment:
Got confused. It actually likely the previous routing node has an older cluster state in which we should override the existing value..  nevermind